tests/unit/test_endpoint_provider.py

```diff
@@ -544,3 +553,141 @@ def test_deeply_nested(self):
+            {
+                "with_topology.rule": yaml.safe_dump(
+                    {"alert": "with_topology", "expr": "up{%%juju_topology%%} == 0"}
+                )
+            }
+        )
+
+        # WHEN relation data is generated
+        self.harness = Harness(provider_charm, meta=provider_charm.metadata_yaml)
+        self.harness.set_leader(True)
+        rel_id = self.harness.add_relation("metrics-endpoint", "prom")
+        self.harness.add_relation_unit(rel_id, "prom/0")
+        self.harness.begin_with_initial_hooks()
+
+        # THEN relation data does not contain juju_unit in expr or labels
+        relation = self.harness.charm.model.get_relation("metrics-endpoint")
+        alert_rules = json.loads(relation.data[self.harness.charm.app].get("alert_rules"))
+        for group in alert_rules["groups"]:
+            with self.subTest(group=group["name"]):
+                for rule in group["rules"]:
+                    with self.subTest(rule=rule["alert"]):
+                        self.assertNotIn("juju_unit", rule["labels"])
+                        self.assertNotIn("juju_unit=", rule["expr"])
+
+                        # These two are to prevent false negatives in case something big changes
+                        self.assertIn("juju_application", rule["labels"])
+                        self.assertIn("juju_application=", rule["expr"])
+
+        self.harness.cleanup()
+
+    @patch_network_get(private_address="192.0.8.2")
+    def test_alert_rules_without_topology(self):
+        """Scenario: Free standing rules shouldn't have any topology whatsoever, including unit."""
+        # GIVEN a provider charm has rule file(s) without a topology stub in the expr
+        provider_charm = provider_with_rules_factory(
+            {"free_standing.rule": yaml.safe_dump({"alert": "free_standing", "expr": "up == 0"})}
+        )
+
+        # WHEN relation data is generated
+        self.harness = Harness(provider_charm, meta=provider_charm.metadata_yaml)
+        self.harness.set_leader(True)
+        rel_id = self.harness.add_relation("metrics-endpoint", "prom")
+        self.harness.add_relation_unit(rel_id, "prom/0")
+        self.harness.begin_with_initial_hooks()
+
+        # THEN relation data does not contain juju_unit in expr or labels
+        relation = self.harness.charm.model.get_relation("metrics-endpoint")
+        alert_rules = json.loads(relation.data[self.harness.charm.app].get("alert_rules"))
+        for group in alert_rules["groups"]:
+            with self.subTest(group=group["name"]):
+                for rule in group["rules"]:
+                    with self.subTest(rule=rule["alert"]):
+                        self.assertNotIn("juju_unit", rule["labels"])
+                        self.assertNotIn("juju_unit=", rule["expr"])
+
+        self.harness.cleanup()
+
+    @patch_network_get(private_address="192.0.8.2")
+    def test_alert_rules_with_hard_coded_unit_label(self):
```

Comment:
If the test name suggested what was being checked for it makes the code self documenting. For example 
` def test_endpoint_provider_does_not_overwrite_hard_coded_unit_label` . The name `test_alert_rules_with_hard_coded_unit_label` only says this test has "some thing" to do with "hard coded unit" labels but does not say what.

Reply:
Agreed.
The `endpoint_provider` prefix might be redundant given the filename.

How about:
```suggestion
    def test_unit_label_is_retained_if_hard_coded(self):
```